### ISSUE TITLE
PEP 746: Constrained types with `Annotated`  and `__supports_type__`

### ISSUE THREAD

**Issue by jorenham** [jorenham/optype#89]
The [PEP](https://peps.python.org/pep-0746/) is currently in draft.
It would be nice to have a collection of predicate types `optype.Check{}`, e.g. 

```python
from optype import CanGe
from typing import Annotated, Literal

class CheckGe[T]:
    def __init__(self, rhs: T, /) -> None:
        self.rhs = rhs

    def __supports_type__(self, lhs: CanGe[T], /) -> bool:
        return lhs >= self.rhs

type IsGe0[T: CanGe[Literal[0]]] = Annotated[T, CheckGe(0)]
type IsGe1[T: CanGe[Literal[1]]] = Annotated[T, CheckGe(1)]
```

For `optype` self this is pretty cool, as it allows for more specific type bounds.
For example, `optype.CanLen[R: int]` could be narrowed to `CanLen[R: IsGe0[int]]` (although in this case, enforcing `0 <= _ <= sys.maxsize` would make it absolutely airtight).
Similarly, the shape type `S@optype.numpy.Array[S, T]` could probably also narrowed from `tuple[int, ...]` to `tuple[IsGe0[int], ...]`, and perhaps it's possible to limit the max shape length (i.e. `ndim`) to `<= 64`.

---

**EDIT**

The PEP was changed so that `__supports_type__` now is a class attribute, instead of a method. This makes way more sense to me, even more so than my earlier idea of using a callable for this purpose.

I believe that this is how the example above should be rewritten as:

```python
from optype import CanGe
from typing import Annotated, Literal

class IsGe[T]:
    __supports_type__: CanGe[T], /)

    def __init__(self, rhs: T, /) -> None:
        self.rhs = rhs

type NonNegativeInt = Annotated[int, IsGe(0)]
type PositiveInt = Annotated[int, IsGe(1)]
```

**Comment by jorenham** [comment]
https://discuss.python.org/t/pep-746-typedmetadata-for-type-checking-of-pep-593-annotated/53834/72

**Comment by jorenham** [comment]
Apparantly this PEP doesn't do what I expected it to do, see https://discuss.python.org/t/pep-746-typedmetadata-for-type-checking-of-pep-593-annotated/53834/75.

This PEP only applies to "the inside" of the `Annotated`, and isn't interpreted as an intersection type from "the outside".

So in the examples above, `NonNegativeInt` and `PositiveInt` are both equivalent to `int`. 

**TLDR;** 
This PEP is useless for `optype`, and doesn't do what you'd expect it to do.
